#!/bin/bash

# author: Andy Tock
# date: 07/02/2022

# After applying mtsv2bedGraph_deepsignal.py (adapted to work with a DeepSignal raw read output TSV;
# original Nanopolish-based script mtsv2bedGraph.py from https://github.com/isaclee/nanopore-methylation-utilities )
# to generate a sorted, bgzipped, and tabix-indexed BED-style format methylation file:
# "Using the converted bed-style methylation file, the original bam file [e.g., generated by winnowmap]
# can be 'bisulfite converted in silico' for easy visualization on IGV via their bisulfite mode."
# Note that this latter step cannot be run with newer version of samtools (> v1.9; e.g., 1.10 in ChIP_mapping conda env)
# due to changes to the SAM format spec

# Usage on node7 (where context=CHH , must be allocated to plantsci cluster node10 as it requires > 300 GB RAM):
# csmit -m 400G -c 48 "bash ./per_read_bam_conversion.sh Col_0_CENH3_ChIP_deepsignalDNAmeth t2t-col.20210610 48 5G CHH" 

prefix=$1
refbase=$2
threads=$3
sortMem=$4
context=$5

source ~/.bashrc

conda activate ChIPseq_mapping

echo $(which tabix)
echo $(which python3)
echo $(which bgzip)

### Generate a sorted, bgzipped, and tabix-indexed BED-style format methylation file
### See https://github.com/isaclee/nanopore-methylation-utilities
### --offset may need to be changed from 0 to 1 for DeepSignal raw read output TSV,
### as coordinates are 0-based, whereas those for the equivalent Nanopolish output are 1-based
### I haven't done this in my tests with mtsv2bedGraph_deepsignal.py as it's not clear whether or not
### the BED-style output of mtsv2bedGraph.py should have 0-based or 1-based start coordinates
### (these should be 0-based if following BED format specs; additionally, mtsv2bedGraph.py adds 1
### to the final cytosine ("read end") coordinate, suggesting that start coordinates are 0-based and
### end coordinates are 1-based)
#zcat ${prefix}_MappedOn_${refbase}_${context}_raw.tsv.gz | grep ${chr} > ${prefix}_MappedOn_${refbase}_${context}_raw_${chr}.tsv   
#python3 mtsv2bedGraph_deepsignal.py --input ${prefix}_MappedOn_${refbase}_${context}_raw_${chr}.tsv \
#                                    --mod ${context} \
#                                    --call-threshold 0.02 \
#                                    --genome /home/ajt200/rds/hpc-work/nanopore/${refbase}/${refbase}.fa \
#                                    --offset 0 \
#                                    --verbose | \
#sort -k1,1 -k2,2n > ${prefix}_MappedOn_${refbase}_${context}_${chr}.bed
cat ${prefix}_MappedOn_${refbase}_${context}_Chr1.bed \
    ${prefix}_MappedOn_${refbase}_${context}_Chr2.bed \
    ${prefix}_MappedOn_${refbase}_${context}_Chr3.bed \
    ${prefix}_MappedOn_${refbase}_${context}_Chr4.bed \
    ${prefix}_MappedOn_${refbase}_${context}_Chr5.bed \
    ${prefix}_MappedOn_${refbase}_${context}_ChrM.bed \
    ${prefix}_MappedOn_${refbase}_${context}_ChrC.bed \
| sort -k1,1 -k2,2n | bgzip > ${prefix}_MappedOn_${refbase}_${context}.bed.gz 
tabix -p bed ${prefix}_MappedOn_${refbase}_${context}.bed.gz
rm ${prefix}_MappedOn_${refbase}_${context}_Chr*.bed
rm ${prefix}_MappedOn_${refbase}_${context}_raw_Chr*.tsv

# "Using the converted bed-style methylation file, the original bam file [e.g., generated by winnowmap]
# can be 'bisulfite converted in silico' for easy visualization on IGV via their bisulfite mode."
# See https://github.com/isaclee/nanopore-methylation-utilities
# Note that this latter step cannot be run with newer version of samtools (> v1.9; e.g., 1.10 in ChIP_mapping conda env)
# due to changes to the SAM format spec
prefix_substr=$(echo ${prefix}_MappedOn_${refbase} | cut -d'_' -f 1-4,6-7)
context_lower=$(echo ${context} | awk '{print tolower($0)}')
python3 convert_bam_for_methylation_deepsignal.py --threads ${threads} \
                                                  --verbose \
                                                  --fasta /home/ajt200/rds/hpc-work/nanopore/${refbase}/${refbase}.fa \
                                                  --bam ${prefix_substr}_winnowmap_sorted.bam \
                                                  --mbed ${prefix}_MappedOn_${refbase}_${context}.bed.gz \
                                                  --mod ${context_lower} | \
  /usr/local/software/spack/spack-0.11.2/opt/spack/linux-rhel7-x86_64/gcc-5.4.0/samtools-1.9-vf6vvemcaplhykbjjketss7he42t6tzr/bin/samtools sort -@ ${threads} -m ${sortMem} -o ${prefix}_MappedOn_${refbase}_${context}_winnowmap_sorted_BSconverted_sorted.bam
/usr/local/software/spack/spack-0.11.2/opt/spack/linux-rhel7-x86_64/gcc-5.4.0/samtools-1.9-vf6vvemcaplhykbjjketss7he42t6tzr/bin/samtools index ${prefix}_MappedOn_${refbase}_${context}_winnowmap_sorted_BSconverted_sorted.bam

conda deactivate
